### DISCUSSION TITLE
halo 2.0 主题问题

### DISCUSSION THREAD

**Discussion by liuzhihang** [halo-dev/halo#2359] [General]
看到九月底发布 halo 2.0，想问下 2.0 版本的主题开发还是使用 freemarker 么？应该是兼容现有主题吧。

因为我想开发主题，考虑到 2.0 改动可能较大，所以咨询一下。

**Comment by JohnNiang** [comment]
Halo 2.0 将采用 [Thymeleaf](https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html) 作为模板引擎。目前我们已经在尝试移植 Anatole 主题，可参考：https://github.com/halo-sigs/theme-anatole

**Reply by liuzhihang** [reply]
那我可以先了解一下 Thymeleaf，期待 2.0 的发布

**Comment by liuzhihang** [comment]
我看到 Anatole 使用的是 Astro 编译之后就可以直接在 halo 2.0 中使用了，并且也可以直接 run 预览，这块后续会出相关开发文档么？或者推荐推荐相关资料。

**Reply by ruibaby** [reply]
可以先看看 https://github.com/halo-sigs/theme-astro-starter 仓库的 README。使用 Astro 编译模板只是主题开发的其中一个 Starter，后续会将开发文档放在 halo-sigs/theme-astro-starter 仓库。

**Reply by liuzhihang** [reply]
感谢。我先看看

**Comment by liuzhihang** [comment]
我看到可以在 Astro 写 vue，并且也支持很多其他框架，这里是不是都能编译为 halo 模版？那我了解一下 vue 就可以了吧

**Reply by ruibaby** [reply]
目前是不行的，建议只用 Vue 做交互相关的组件，因为数据交互还是通过我们提供的模板引擎方法。所以你可能还是需要熟悉 Thymeleaf 语法以及我们提供的主题 API。当然，我们也提供了 HTTP API，你也可以用 API 进行数据交互。

**Reply by liuzhihang** [reply]
好的，感谢，我现在先了解一下相关框架，等 2.0 发布之后，参考文档再试试自己开发主题。

**Comment by liuzhihang** [comment]
@JohnNiang @ruibaby  咱们月底发布 halo 2.0，基于 Astro 开发主题的 开发文档是不是也在月底之前出来

**Reply by ruibaby** [reply]
会提供主题开发快速开始、配置、主题端 API 文档。配合 Astro 开发的文档可能会更新到 https://github.com/halo-sigs/theme-astro-starter

**Reply by JohnNiang** [reply]
可能没有足够的时间提供详细的文档，不过简单的开始会有的。

**Comment by liuzhihang** [comment]
@JohnNiang @ruibaby  问个问题。
如果完全使用 React 开发前端，调用 API 。这样的话就不能在 admin 管理这个前后端分离的主题了吧。就是像现在 admin 中一样快速切换主题，设置主题的各种自定义内容、页面等。还得自己单独部署前端，会比较麻烦。

所以还是推荐使用 thymeleaf 或者 Astro + React 组件或者 Vue 组件这种来搞吧？这样就可以在  admin 中提供给其他小伙伴来使用这个主题了。

不知道这么理解对不对。

**Reply by guqing** [reply]
> @JohnNiang @ruibaby 问个问题。 如果完全使用 React 开发前端，调用 API 。这样的话就不能在 admin 管理这个前后端分离的主题了吧。就是像现在 admin 中一样快速切换主题，设置主题的各种自定义内容、页面等。还得自己单独部署前端，会比较麻烦。
> 
> 所以还是推荐使用 thymeleaf 或者 Astro + React 组件或者 Vue 组件这种来搞吧？这样就可以在 admin 中提供给其他小伙伴来使用这个主题了。
> 
> 不知道这么理解对不对。

企业级理解，👍

**Reply by JohnNiang** [reply]
可以这样理解。小程序就是一个活生生的例子。

**Reply by guqing** [reply]
使用 react 用法应该和 Astro+Vue 差不多吧 @ruibaby ,只要写 html 的时候要加上 thymeleaf 模板语法 并使其 build 后在 templates 目录输出正确的主题结构即可

**Reply by ruibaby** [reply]
如果是使用 React 或 Vue 编写单页面主题，也是可以通过后台来管理的。只需要配置构建工具将输出的构建资源放在 `path/to/{theme}/templates` 即可。当然一般构建这样的单页面的话，路由基本就只能交给前端来处理了，也就是只有一个 index.html 模板，路由切换交给前端。那么这时候最好就是使用 `hash route` 了，或者配置 Nginx 将所有路由都转发到 /。

主题项目的 `theme.yaml` `settings.yaml` 该怎么放就怎么放了，后续也会有获取主题配置的接口。

**Reply by ruibaby** [reply]
> 用 react 用法应该和 Astro+Vue 差不多吧 @ruibaby ,只要写 html 的时候要加上 thymeleaf 模板语法 并使其 build 后在 templates 目录输出正确的主题结构即可

不行的，React 单页应用都是由客户端渲染的，几乎不能使用我们为 Thymeleaf 提供的主题 API。

**Reply by guqing** [reply]
> > 用 react 用法应该和 Astro+Vue 差不多吧 @ruibaby ,只要写 html 的时候要加上 thymeleaf 模板语法 并使其 build 后在 templates 目录输出正确的主题结构即可
> 
> 不行的，React 单页应用都是由客户端渲染的，几乎不能使用我们为 Thymeleaf 提供的主题 API。

 这里的 react 是指 Astro+React

**Reply by ruibaby** [reply]
> 这里的 react 是指 Astro+React

也是不能在 React 组件里面使用 Thymeleaf 的主题 API 的，因为组件会被编译为 JavaScript。

**Reply by liuzhihang** [reply]
所以可以使用 Astro + Vue 或者 纯粹使用 thymeleaf么，还是说 vue也存在这个问题？
@ruibaby
